### ISSUE TITLE
Two notes on Querying perdomrance are lacking background information

### ISSUE THREAD

**Issue by marwalsch** [dotnet/EntityFramework.Docs#3420]
This issue involves two notes on the mentioned page.

> Avoid using ToList or ToArray if you intend to use another LINQ operator on the result - this will needlessly buffer all results into memory. Use AsEnumerable instead.

Excuse me if I am misinformed here, but does `AsEnumerable()` not include the results in memory as well? I thought the mentioned distinction corresponds to `IEnumerable `and `IQueryable` only. 

> Avoid mixing synchronous and asynchronous code in the same application - it's very easy to inadvertently trigger subtle thread-pool starvation issues.

Since synchronous code results in blocking calls I can imagine that allocated threads for asynchronous methods might lead to starvation in certain scenarios, but isn't mixing it up still a better solution than going fully synchronous?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 4e33b6d9-4d54-138e-7ba2-a1cfbda8b233
* Version Independent ID: a079c83a-fa76-efb6-e744-d01177379fce
* Content: [Efficient Querying - EF Core](https://docs.microsoft.com/en-us/ef/core/performance/efficient-querying)
* Content Source: [entity-framework/core/performance/efficient-querying.md](https://github.com/dotnet/EntityFramework.Docs/blob/main/entity-framework/core/performance/efficient-querying.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @roji
* Microsoft Alias: **avickers**

**Comment by roji** [comment]
> Excuse me if I am misinformed here, but does AsEnumerable() not include the results in memory as well?

No, it does not. AsEnumerable causes the query to be executed, but streams the results back without buffering them in memory. That is what the IEnumerable abstraction is generally about.

> Since synchronous code results in blocking calls I can imagine that allocated threads for asynchronous methods might lead to starvation in certain scenarios, but isn't mixing it up still a better solution than going fully synchronous?

There are some particularly subtle and dangerous issues that can occur when mixing sync and async which do not occur when doing fully synchronous I/O. For example, imagine a Winforms application where there's a single UI thread (and a special SynchronizationContext which schedules continuations to that thread). If you perform sync I/O on that thread (i.e. in the callback of some UI button), your UI will indeed freeze for the time it takes the I/O to complete, but that's about it (so UI responsiveness issue). If, however, you synchronously block on some async operation, you deadlock the application - [see this article for more details](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html).

To summarize, if you're careful and know what you're doing, it's possible to mix sync and async, but it's definitely not trivial.

**Comment by marwalsch** [comment]
@roji  Thanks a lot.
